enhancements/ingress/add-headers-to-haproxyconfig-via-ingress-controller-or-route-api.md

```diff
@@ -0,0 +1,413 @@
+
+* openshift/api
+* openshift/cluster-ingress-operator
+* openshift/router
+
+The router configures HAProxy using a configuration template.  The template uses
+environment variables as input parameters.  The enhancement adds a new
+environment variables, `ROUTER_SET_HTTP_HEADERS`, which specify the headers to set in the
+frontend fe_no_sni, frontend public, frontend fe_sni  section of the haproxy.config.
+The ingress operator sets this variable based on the IngressController's
+`spec.httpSetHeaders` field value.
+
+The values of `ROUTER_SET_HTTP_HEADERS` are comma-delimited lists of base64 encoded values of
+the form `name:value` separated by comma.
+The template translates these values into an appropriate `http-response set-header` stanzas.
+For example, if `ROUTER_SET_HTTP_HEADERS` has encoded value `WC1GcmFtZS1PcHRpb25zOiBERU5Z,WC1YU1MtUHJvdGVjdGlvbjogMTttb2RlPWJsb2Nr` 
+of `X-Frame-Options: DENY,X-XSS-Protection: 1;mode=block` then the template adds the following stanzas to
+the frontend fe_no_sni, frontend public, frontend fe_sni  section of the haproxy.config:
+```conf
+    http-response set-header X-Frame-Options: DENY
+    http-response set-header X-XSS-Protection: 1;mode=block    
+```
+
+This feature's implementation will be similar to an existing feature called logging headers which was implemented in 
+[router](https://github.com/openshift/router/commit/40ee0ba472a4ac503f1e6c97d4464f559101c97c) and 
+[cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/commit/c5139cb07c3a99a8a3b18a6c9321310add51b075).
+
+B) Setting HTTP headers via Route's Spec by Users/Devops/Developers.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/router
+
+The router configures HAProxy using a configuration template. The template is populated with values in ServiceAliasConfig .
+The enhancement adds a field in ServiceAliasConfig called `SetHTTPHeadersViaRoute`, which specify the headers to set in the
+backend section `Plain http backend or backend with TLS terminated at the edge or a secure backend with re-encryption.` of the
+haproxy.config at this [code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L609).
+The router sets this variable based on the Route's
+`spec.httpSetHeaders` field value. Service AliasConfig will be populated at this [code location](https://github.com/openshift/router/blob/master/pkg/router/template/router.go#L927)
+
+The values of `SetHTTPHeadersViaRoute` will be set in the same way as described in the section A) above for  `ROUTER_SET_HTTP_HEADERS`
+
+### Risks and Mitigations
+
+### Drawbacks
+
+## Design Details
+
+### Open Questions [optional]
+
+- Do we need to set request headers ?
+- Do we need a delete option ?
+- Do we a need admission list option ?
+- Do we need to set load balance algo based on the header set ?
```

Comment:
I have no idea why we would set the balance algorithm based on an HTTP header.  Where does this question come from?  Is there a particular header you have in mind?  Is this question related to some statement or comment in the RFE, or some request someone has conveyed to you?

Reply:
someone in the jira ticket was asking for it.

Reply:
I see, the request is that if a route is annotation with `haproxy.router.openshift.io/balance: hdr(x-foo)`, then OpenShift router should write `balance hdr(x-foo)` in `haproxy.config`.  That's trivial to do, but it's orthogonal to customizing HTTP headers.  I agree that it is out of scope for this enhancement and should be handled with a separate RFE.

Reply:
`Setting load balancing algorithm based on headers` is already mentioned in the `Non-goals`. Doesn't need to be in the open question. As a matter of fact I think this doesn't need to be mentioned at all since:
1. We seem to go with the response only API while `haproxy.router.openshift.io/balance: hdr()`makes sense for the requests only.
2. Even if we'd have the request API too that would only be a helper for the `haproxy.router.openshift.io/balance: hdr()` annotation, not enabler or prerequisite.

Reply:
It makes sense to me to list it as a non-goal because someone asked for it on the RFE.  I don't know why the person who asked for it thought a feature for customizing headers would include "balance traffic based on the header", but since someone _did think_ that, we might as well state explicitly that the enhancement _does not_ include it.  

Anyway, I think we are all in agreement that this is a non-goal, so @miheer, please add some text below this question in the EP to say that the question has been resolved, and that we have decided that this is a non-goal for this EP.

Reply:
@miheer, reminder to add some text to the EP for this (and other) open questions.  I'll also provide some examples in case that's helpful.  

We have several DNS- and ingress-related EPs that present open questions (or questions that were previously open) in a Q&A style, such as the [forward-dns-over-tls](https://github.com/openshift/enhancements/blob/master/enhancements/dns/forward-dns-over-tls.md#open-questions-optional), [cache-tuning](https://github.com/openshift/enhancements/blob/master/enhancements/dns/cache-tuning.md#open-questions-optional), and [lb-allowed-source-ranges](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/lb-allowed-source-ranges.md#open-questions-optional) EPs.  [The ip-interface-selection EP](https://github.com/openshift/enhancements/blob/master/enhancements/network/ip-interface-selection.md#open-questions-optional) has a similar format, and [the configurable-dns-loadbalancerservice EP](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/configurable-dns-loadbalancerservice.md#open-questions) similarly has an open question that has been "closed" with an explanation.  We sometimes put open questions in the EP for follow-up, for example [the route-subdomain EP](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/route-subdomain.md#open-questions).  We sometimes write open questions with a note to document the issue, as in the [haproxy-healthcheckinterval](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/haproxy-healthcheckinterval.md#open-questions) and [tunable-router-kubelet-probes](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/tunable-router-kubelet-probes.md#open-questions) EPs.  We also sometimes have an open question with a link to an issue tracker, as with [the contour-operator EP](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/contour-operator.md#open-questions-optional).  The console [customize-add-page](https://github.com/openshift/enhancements/blob/master/enhancements/console/customize-add-page.md#open-questions) and [customize-project-roles](https://github.com/openshift/enhancements/blob/master/enhancements/console/customize-project-roles.md#open-questions) EPs do a really nice job of explaining the question with an example and quoting the decision that was made on the question (the only thing I would add would be a citation for the answer).  [The infrastructure-external-platform-type EP](https://github.com/openshift/enhancements/blob/master/enhancements/cloud-integration/infrastructure-external-platform-type.md#open-questions) does a really nice job of listing open questions and "answered" questions with detailed answers and citations.  

The purpose of writing down these open questions is to show some of the thought that went into the design so that people can read that these questions have been raised, discussed, and resolved (or possibly that they are still open questions that may warrant follow-up).  

The next person who reads the EP cannot be expected to read through all the GitHub comments; it's your job to make sure that the EP itself has this information.

Reply:
done